src/state/slices/opportunitiesSlice/selectors/combined.ts

```diff
@@ -297,12 +302,82 @@ export const selectAggregatedEarnOpportunitiesByProvider = createDeepEqualOutput
     includeEarnBalances,
     includeRewardsBalances,
     chainId,
+    searchQuery,
   ): AggregatedOpportunitiesByProviderReturn[] => {
     if (isEmpty(marketData)) return []
     const totalFiatAmountByProvider = {} as Record<DefiProvider, BN>
     const projectedAnnualizedYieldByProvider = {} as Record<DefiProvider, BN>
     const combined = [...userStakingOpportunites, ...userLpOpportunities]
 
+    /**
+     * we want to be able to search on...
+     * - provider "thorch" for THORChain, "unis" for Uniswap
+     * - asset name (vault/underlying/rewards) e.g. idle 'senior tr' for Idle Senior TR, 'usdc' will match too
+     * - chain "opt" for Optimism
+     *
+     * https://github.com/kentcdodds/match-sorter#advanced-options
+     *
+     * we are using the function style advanced filtering of match-sorter to map
+     * - assetId -> asset name
+     * - chainId -> chain name
+     * - rewardAssetIds[] - asset name[]
+     * - underlyingAssetIds[] - asset name[]
+     *
+     * - if we include a search term, we want to match on any of these
+     * - if we don't include a search term, return all
+     *
+     * the search should resolve more broadly than narrowly, e.g.
+     * i search for "compound" - i should see all compound vaults, not just the one i'm looking for
+     */
+
+    const searchOpportunities = <T extends LpEarnOpportunityType | StakingEarnOpportunityType>(
```

Comment:
There's too many of all sorts of brackets including SQUIRLY BRECKITS for my brain to grasp this but I trust GPT-4 and will give this a thorough testing